src/expr/core/src/window_function/session.rs

```diff
@@ -0,0 +1,208 @@
+            f,
+            "SESSION GAPPED BY {}",
+            self.gap.as_scalar_ref_impl().to_text()
+        )
+    }
+}
+
+impl FrameBoundsImpl for SessionFrameBounds {
+    fn validate(&self) -> Result<()> {
+        // TODO(rc): maybe can merge with `RangeFrameBounds::validate`
+
+        fn validate_non_negative(val: impl IsNegative + Display) -> Result<()> {
+            if val.is_negative() {
+                bail!("session gap should be non-negative, but {} is given", val);
+            }
+            Ok(())
+        }
+
+        match self.gap.as_scalar_ref_impl() {
+            ScalarRefImpl::Int16(val) => validate_non_negative(val)?,
+            ScalarRefImpl::Int32(val) => validate_non_negative(val)?,
+            ScalarRefImpl::Int64(val) => validate_non_negative(val)?,
+            ScalarRefImpl::Float32(val) => validate_non_negative(val)?,
+            ScalarRefImpl::Float64(val) => validate_non_negative(val)?,
+            ScalarRefImpl::Decimal(val) => validate_non_negative(val)?,
+            ScalarRefImpl::Interval(val) => {
+                if !val.is_never_negative() {
+                    bail!(
+                        "for session gap of type `interval`, each field should be non-negative, but {} is given",
+                        val
+                    );
+                }
+                if matches!(self.order_data_type, DataType::Timestamptz) {
+                    // for `timestamptz`, we only support gap without `month` and `day` fields
+                    if val.months() != 0 || val.days() != 0 {
+                        bail!(
+                            "for session order column of type `timestamptz`, gap should not have non-zero `month` and `day`",
+                        );
+                    }
+                }
+            }
+            _ => unreachable!(
+                "other order column data types are not supported and should be banned in frontend"
+            ),
+        }
+        Ok(())
+    }
+}
+
+impl SessionFrameBounds {
+    pub fn minimal_next_start_of(&self, end_order_value: impl ToOwnedDatum) -> Datum {
+        self.gap.for_calc().minimal_next_start_of(end_order_value)
+    }
+}
+
+/// The wrapper type for [`ScalarImpl`] session gap, containing an expression to help adding the gap
+/// to a given value.
+#[derive(Debug, Clone, Educe)]
+#[educe(PartialEq, Eq, Hash)]
+pub struct SessionFrameGap {
+    /// The original gap value.
+    gap: ScalarImpl,
+    /// Built expression for `$0 + gap`.
+    #[educe(PartialEq(ignore), Hash(ignore))]
+    add_expr: Option<Arc<BoxedExpression>>,
+}
+
+impl Deref for SessionFrameGap {
+    type Target = ScalarImpl;
+
+    fn deref(&self) -> &Self::Target {
+        &self.gap
+    }
+}
+
+impl SessionFrameGap {
+    pub fn new(gap: ScalarImpl) -> Self {
+        Self {
+            gap,
+            add_expr: None,
+        }
+    }
+
+    fn prepare(&mut self, order_data_type: &DataType, gap_data_type: &DataType) -> Result<()> {
```

Comment:
I separate it because it's only need to be called when backend restore the structure from protobuf. When newing the struct in frontend, the `add_expr` inside is not used.

Reply:
Oh, thanks, it came back to me that the range frame also has similar logic 👍